code/modules/projectiles/projectile.dm

```diff
@@ -404,10 +408,12 @@
 	trajectory.increment(trajectory_multiplier)
 	var/turf/T = trajectory.return_turf()
 	if(T.z != loc.z)
+		var/old = loc
 		before_z_change(loc, T)
 		trajectory_ignore_forcemove = TRUE
 		forceMove(T)
 		trajectory_ignore_forcemove = FALSE
+		after_z_change(old, loc)

```

Comment:
Whatever you're doing here, could you do it in a `onTransitZ(old_z, new_z)` proc? And if you can, please be sure to call parent.

Reply:
@vuonojenmustaturska it's needed because transits aren't the only thing that changes zlevels

Reply:
onTransitZ is supposed to fire for all Z-level transitions. It has some exceptions, like items inside ore boxes or vending machines, and ghost/camera mobs, but everything else is probably a bug. However, you only get references to Z levels with it.